packages/php-wasm/scopes/src/index.ts

```diff
@@ -115,3 +136,35 @@ export function removeURLScope(url: URL): URL {
 	newUrl.pathname = '/' + parts.slice(2).join('/');
 	return newUrl;
 }
+
+/**
+ * Appends a pathname to the base URL and maintains scope information.
+ *
+ * If the pathname is scoped, it will maintain scope information from the pathname.
+ * Otherwise, the pathname of the the base URL pathname to preserve the scope.
+ *
+ * If neither the pathname nor the base URL have scope information,
+ * the base URL will be prepended to the pathname and the result will be unscoped.
+ *
+ * @example
+ * ```js
+ * appendPathnameToBaseUrlAndMaintainScope(new URL('http://localhost/scope:base/'), '/scope:pathname/index.php');
+ * // 'http://localhost/scope:pathname/index.php'
+ *
+ * appendPathnameToBaseUrlAndMaintainScope(new URL('http://localhost/scope:base/', '/index.php'));
+ * // 'http://localhost/scope:base/index.php'
+ * ```
+ *
+ * @param  baseUrl  - The base URL to prepend to the pathname.
+ * @param  pathname - The pathname to append to the base URL.
+ * @returns The absolute URL.
+ */
+export function appendPathnameToBaseUrlAndMaintainScope(
+	baseUrl: URL,
+	pathname: string
+): string {
+	if (isUrlPathnameScoped(pathname)) {
+		return baseUrl.origin + pathname;
+	}
+	return baseUrl.origin + joinPaths(baseUrl.pathname, pathname);
```

Comment:
I'm not sure whether or not these concatenations are safe with regard to leading and trailing slashes. For example, this is true in Safari:

```js
((new URL("http://example.com")).origin + "xyz") === "http://example.comxyz"
```

To more easily deal with edge cases around leading slashes, we can use the URL constructor like:

```ts
(new URL("xyz", "http://example.com").toString()) === "http://example.com/xyz"
```

Reply:
Still, what’s the use of this beyond the request handler? It feels like an API added just in case it’s needed. Are there actual multiple use-cases? It is a complex function and it’s not clear what to do when, say, it sees two mismatched scopes.

Reply:
If I understand this correctly, you are saying, this is specific to the 300 redirect code and it should live there if it's needed and not as a helper function?

The code is removed now, but I would like to understand it for the future.

Reply:
All I’m saying is I don’t see a use-case for merging an absolute siteURL that contains a scope with a path that also contains a scope.